File: core/cloudtts/src/main/kotlin/xyz/libravault/core/cloudtts/CloudTtsProvider.kt
package xyz.libravault.core.cloudtts

/**
 * Vendor-agnostic cloud TTS synthesis, dispatched by [CloudProviderId].
 *
 * Real implementation ([xyz.libravault.core.cloudtts.RealCloudTtsProvider],
 * `src/play` — added in the vendor-adapters follow-up) fans out to one of the
 * five vendor HTTP adapters. F-Droid gets a `NoOpCloudTtsProvider` (`src/fdroid`)
 * that fails every call, matching [xyz.libravault.core.billing.SupportBillingClient]
 * / `NoOpBillingClient`'s shape — this module never pulls networking
 * dependencies into the F-Droid build.
 *
 * Neither method is reachable unless [CloudTtsGate.observeCanUseCloudTts]
 * currently reports true — callers (this module's own `CloudTtsEngine`,
 * added in the engine-wiring follow-up) are responsible for checking the
 * gate immediately before every call, not just once.
 */
interface CloudTtsProvider {

    /** Synthesizes [text] as [voiceId] using [apiKey], returning encoded audio
     * bytes on success. Non-streaming, per-paragraph — PRD §3. */
    suspend fun synthesize(provider: CloudProviderId, text: String, voiceId: String, apiKey: String): Result<ByteArray>

    /** A cheap call that validates [apiKey] is accepted by [provider] without
     * synthesizing real audio — PRD §6, shown as immediate feedback when the
     * user pastes a key into Settings. */
    suspend fun validateKey(provider: CloudProviderId, apiKey: String): Result<Unit>
}

File: ios/LibraVaultApp/LibraVault/LibraVault/Sources/Features/CloudTTS/CloudTtsProvider.swift
import Foundation

/// One HTTP call shape shared by all five vendor adapters (`Vendors/*Adapter.swift`) —
/// iOS counterpart to Android's `CloudTtsProvider` interface (core/cloudtts/CloudTtsProvider.kt).
/// A protocol, not a direct `URLSession` call at every call site, both for fakeability in
/// tests and so `CloudTtsEngine` can dispatch to whichever vendor is actually selected
/// without a hardcoded `switch` of its own — `RealCloudTtsProvider` (added in the
/// vendors PR) is the dispatcher; this is just the shared contract.
///
/// Every conformer/caller MUST check `CloudTtsGate.canUseCloudTts` immediately before
/// calling either method below — this protocol has no way to enforce that itself, since
/// enforcing it here would require this protocol to depend on `StoreKitBillingManager`/
/// `CloudVoicePreferences`, which would invert the dependency direction this feature is
/// built on (see the implementation plan's "Engine integration seam"). `CloudTtsEngine`
/// (added in the engine PR) is the actual enforcement point.
protocol CloudTtsProvider {
    /// Synthesizes `text` as speech using `voiceID` and `credentials`, returning the raw
    /// audio bytes. Format is vendor-specific (MP3 for most) — handled generically by
    /// `AVAudioPlayer` on the `CloudTtsEngine` side, the same "don't care about the
    /// container format, just play the bytes" approach Android's `MediaPlayer`-based
    /// `CloudPlayback` takes.
    func synthesize(
        provider: CloudProviderId,
        text: String,
        voiceID: String,
        credentials: [CloudCredentialField: String]
    ) async throws -> Data

    /// A single cheap validation call (PRD §6: "validated with a single cheap test
    /// call, then stored") — never persists anything itself. The caller only saves
    /// credentials (via `CloudApiKeyStore`) once this succeeds.
    func validateKey(
        provider: CloudProviderId,
        credentials: [CloudCredentialField: String]
    ) async throws
}

/// Errors this protocol's conformers throw — deliberately generic (not
/// per-vendor-specific HTTP status parsing) since callers only need to know "did it
/// work," not why, to decide whether to fall back to the on-device engine. Mirrors
/// Android's `CloudTtsProvider`'s `Result<...>` failure shape, adapted to Swift's
/// `throws` convention.
enum CloudTtsProviderError: LocalizedError, Equatable {
    case missingCredentials(field: CloudCredentialField)
    case httpError(statusCode: Int, body: String)
    case invalidResponse

    var errorDescription: String? {
        switch self {
        case .missingCredentials(let field):
            return "Missing required field: \(field.label)"
        case .httpError(let statusCode, let body):
            return "HTTP \(statusCode): \(body.prefix(200))"
        case .invalidResponse:
            return "Unexpected response from the TTS vendor"
        }
    }
}
